src/ray/gcs/object_state_accessor.h

```diff
@@ -0,0 +1,88 @@
+#ifndef RAY_GCS_OBJECT_STATE_ACCESSOR_H
+#define RAY_GCS_OBJECT_STATE_ACCESSOR_H
+
+#include "ray/common/id.h"
+#include "ray/gcs/callback.h"
+#include "ray/gcs/subscription_executor.h"
+#include "ray/gcs/tables.h"
+
+namespace ray {
+
+namespace gcs {
+
+class RedisGcsClient;
+
+/// \class ObjectStateAccessor
+/// ObjectStateAccessor class encapsulates the implementation details of
+/// reading or writing or subscribing of object's information.
+class ObjectStateAccessor {
+ public:
+  explicit ObjectStateAccessor(RedisGcsClient &client_impl);
+
+  ~ObjectStateAccessor() {}
+
+  /// Get object information from GCS asynchronously.
+  ///
+  /// \param object_id The ID of object to lookup in GCS.
+  /// \param callback Callback that will be called after lookup finishes.
+  /// \return Status
+  Status AsyncGet(const ObjectID &object_id,
```

Comment:
Could change to `AsyncLookupLocations(object_id, MultiItemCallback<NodeId> callback)` to make this interface more focused on the semantics, rather than implementation.